AzureAD/azure-activedirectory-library-for-dotnet
# Windows Integrated Auth (Approved, do not merge until we ship next release)

ADAL dotnet has long been supporting Windows Integrated Auth (WIA) feature. MSAL dotnet does not. This PR implements WIA infrastructure in the `core` code base which is shared by both ADAL and MSAL, and then implement the WIA methods in MSAL, and also refactor WIA in ADAL to use those shared infrastructure in `core`, so that the future maintenance would only need to deal with one set of code, in `core`.

To review this PR, you can either read through those thousands lines of additions as a whole, or if you prefer, you can review them in "bite-size", by each commit in the history of this PR. Those commits have already been reorganized to reflect a meaningful development history.

Some things worth mentioning here:

* We changed some pre-existing ADAL test cases to reflect the fact that the relevant functionality relies on the new components from core now. But we hardly changed any asserts, therefore all the expected behaviors remain the same, and the new WIA infrastructure in core has same level of tests coverage and confidence, tested by ADAL.
* We did change 2 occurrences of `Assert(...)` that, previously a generic `Exception` was expected, now the new code becomes more explicit and ensures the proper `AdalException` would be raised.
* There are several commented-out code blocks `/* ... */` inside this PR. They are not strickly part of this PR but they would be a good starting point for the upcoming Username/Password flow implementation. I leave those commented blocks here, for now.
* The new WIA methods in MSAL are currently implemented on all platforms. But I guess we would only want it to be available on Windows? We can adjust that.

Steps for manual test:

1. Assuming you are going to use your corp account to test, so you will want to do the follow steps on a domain-joined desktop (or laptop);
2. Launch the MSAL Desktop Test App, change the default authority to `https://login.microsoftonline.com/organizations`. (This is a requirement by AAD. You can skip this step and see the corresponding error message firsthand.)
3. Hit the "Acquire Token By WindowsIntegratedAuth" button. The request will be sent out, and you should see the response.
4. You may, however, encounter this error message "AADSTS50079: The user is required to use multi-factor authentication". See an explanation [here](https://stackoverflow.com/questions/45799868/login-to-azure-web-application-fails-with-aadsts50079-the-user-is-required-to). The solution is to trigger an interactive flow, by somehow construct and visit a url like this (but replace the tenant and client_id with yours): `https://login.microsoftonline.com/microsoft.onmicrosoft.com/oauth2/v2.0/authorize?client_id=0615b6ca-88d4-4884-8729-b178178f7c27&response_type=code&redirect_uri=http%3A%2F%2Flocalhost%3A44331&scope=user.read email offline_access openid profile`.
   This solution may still not solve that error, though. If the tenant's admin configures an MFA policy, you can not use WIA to acquire token. Other than that, this PR already does all the WIA flow and send the correct request to AAD.
